aten/src/ATen/native/cpu/SumKernel.cpp

```diff
@@ -452,16 +585,94 @@ void sum_kernel_impl(TensorIterator &iter) {
           const int64_t out_stride = out_strides[1];
           TORCH_INTERNAL_ASSERT(out_strides[0] == 0);
 
-          if (in_strides[0] == sizeof(scalar_t) && size0 >= Vectorized<scalar_t>::size()) {
+          using vec_t = Vectorized<scalar_t>;
+          using acc_t = at::acc_type<scalar_t, true>;
+          using vacc_t = Vectorized<acc_t>;
+          using ScalarLoadPolicy = CastLoadPolicy<scalar_t, acc_t>;
+          using StorePolicy = CastStoreAccumulate<scalar_t, acc_t>;
+
+          if (in_strides[0] == sizeof(scalar_t) && size0 >= vec_t::size()) {
+            // Contiguous inner reduction
+            using VecLoadPolicy = InnerSumCastLoadPolicy<vec_t, vacc_t>;
+            vectorized_inner_sum<acc_t, VecLoadPolicy, ScalarLoadPolicy, StorePolicy>(
+                data, in_strides[1], out_stride, size0, size1);
+          } else if (in_strides[1] == sizeof(scalar_t) && size1 >= vec_t::size()) {
+            // Contiguous outer reduction
+            using VecLoadPolicy = OuterSumCastLoadPolicy<vec_t, vacc_t>;
+            vectorized_outer_sum<acc_t, VecLoadPolicy, ScalarLoadPolicy, StorePolicy>(
+                data, in_strides[0], out_stride, size0, size1);
+          } else if (in_strides[0] < in_strides[1]) {
+            scalar_inner_sum<acc_t, ScalarLoadPolicy, StorePolicy>(
+                data, in_strides, out_stride, size0, size1);
+          } else {
+            scalar_outer_sum<acc_t, ScalarLoadPolicy, StorePolicy>(
+                data, in_strides, out_stride, size0, size1);
+          }
+        });
+    });
+}
+
+void nansum_kernel_impl(TensorIterator &iter) {
```

Comment:
similarly, this looks very similar to regular sum kernel impls, so maybe it could be unified also? (If it's tricky for some reason that I don't see, it's fine, but having repeated non-trivial code makes me nervous)

Reply:
The "not a reduction" loop can't be shared because of the extra `isnan` checks and all of the `foo_sum` calls are done with different template parameters from `sum_kernel_impl`. That leaves the only common code as swapping strides and some minor boiler plate associated with `basic_loop`.

Reply:
Great refactor with the `cascade_sum` helper.